test/index.test.js

```diff
@@ -0,0 +1,243 @@
+const expect = require("chai").expect
+const proxyquire = require("proxyquire")
+const resultFixture = require("./support/fixtures/results")
+
+describe("main", () => {
+  describe(".run(options)", () => {
+    const consoleLogOriginal = console.log
+    after(() => {
+      console.log = consoleLogOriginal
+    })
+
+    const config = {}
+
+    let Analytics
+    let result
+    let PostgresPublisher
+    let ResultFormatter
+    let S3Publisher
+    let main
+
+    beforeEach(() => {
+      result = {}
+      Analytics = {
+        reports: [{ name: "a" }, { name: "b" }, { name: "c" }],
+        query: (report) => Promise.resolve(Object.assign(result, { name: report.name })),
+      }
+      PostgresPublisher = {}
+      ResultFormatter = {
+        formatResult: (result) => Promise.resolve(JSON.stringify(result))
+      }
+      S3Publisher = {}
+
+      main = proxyquire("../index.js", {
+        "./src/config": config,
+        "./src/analytics": Analytics,
+        "./src/publish/postgres": PostgresPublisher,
+        "./src/process-results/result-formatter": ResultFormatter,
+        "./src/publish/s3": S3Publisher,
+      })
+    })
+
+    it("should query for every single report", done => {
+      const queriedReportNames = []
+
+      Analytics.query = (report) => {
+        queriedReportNames.push(report.name)
+        return Promise.resolve(result)
+      }
+
+      main.run().then(() => {
+        expect(queriedReportNames).to.include.members(["a", "b", "c"])
+        done()
+      }).catch(done)
+    })
+
+    it("should log formatted results", done => {
+      ResultFormatter.formatResult = () => Promise.resolve("I'm the results!")
+
+      let consoleLogCalled = false
+      console.log = function(output) {
+        if (output === "I'm the results!") {
+          consoleLogCalled = true
+        } else {
+          consoleLogOriginal.apply(this, arguments)
+        }
+      }
+
+      main.run().then(() => {
+        console.log = consoleLogOriginal
+        expect(consoleLogCalled).to.be.true
+        done()
+      }).catch(err => {
+        console.log = consoleLogOriginal
+        done(err)
+      })
+    })
+
+    it("should format the results with the format set to JSON", done => {
+      let formatResultCalled = false
+      ResultFormatter.formatResult = (result, options) => {
+        expect(options.format).to.equal("json")
+        formatResultCalled = true
+        return Promise.resolve("")
+      }
+
+      main.run().then(() => {
+        expect(formatResultCalled).to.be.true
+        done()
+      }).catch(done)
+    })
+
+    context("with --output option", () => {
+      xit("should write the results to the given path folder")
```

Comment:
Note that I intentionally left this pending b/c I have plans to move the code it would test into a `DiskPublisher` service and want to wait until that is built to mock out the `fs` module.